# 1.2 IC Data Consistency Checks

## Contents
### Import libraries and data sets
### Missing values in products df
### Duplicates in products df
### Export cleaned products data set
### Data consistency checks on the orders data set
### Mixed type data in orders df
### Missing values in orders df
### Duplicates in orders df
### Export cleaned orders df


## Import libraries and data sets

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# create path to project folder
path = r'/Users/susanwang/Documents/CF_Tasks/Instacart Basket Analysis'

In [3]:
# import products data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
# import orders data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

In [6]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Missing Values

In [7]:
# check df_prods for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# create dataframe with only null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [9]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [10]:
# df_prods shape to check rows
df_prods.shape

(49693, 5)

In [11]:
# create clean df dropping nan rows
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [12]:
# check clean df shape
df_prods_clean.shape

(49677, 5)

## Duplicates

In [13]:
# create new df with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# check df shape rows
df_prods_clean.shape

(49677, 5)

In [16]:
# drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
#check new df shape
df_prods_clean_no_dups.shape

(49672, 5)

## Export cleaned products dataset

In [18]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index=False)

## Data consistency checks on the orders dataset

In [19]:
# df_ords describe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_the_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06            3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01            2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01            2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00            0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00            1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01            3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01            5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02            6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

The 'days_since_prior_order' column has less count than the other columns. Are there NaN values there?
The minimums look fine. For the 'day_of_the_week', 0 means Saturday. For hour_of_day, it would be midnight. 
The maximum order_number is 100, but the 25% is 5, 50% is 11, and 75% is 23. We should check the order_number column.

## Mixed type data in orders df

In [20]:
# use for loop, applymap deprecated, using map instead
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)
    else: 
        print(f'all one type for {col}')


all one type for order_id
all one type for user_id
all one type for order_number
all one type for orders_day_of_the_week
all one type for order_hour_of_day
all one type for days_since_prior_order


No mixed data found.

In [21]:
# checking dtypes
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

Order_id and user_id shouldn't be integers. Will need to go back and redo the wrangling step:

In [22]:
# change data type for order_id
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [23]:
# change data type for user_id
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [24]:
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## Missing values in orders df

In [25]:
# finding using isnull
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in days_since_prior_order. That is the same number as the maximum in the user_id column.
Every customer must make his first order, but there can't be a calulation between that first order and his prior order, so that is why the value is missing. 

In [26]:
# address missing values by creating a new variable
df_ords['first_order'] = df_ords['days_since_prior_order'].isnull()

There are too many missing values in this column, so removing the rows is not a good idea.

I cannot impute the average value, because that would distort the data. We want to keep the information about the customer's first order.
I also cannot impute a representative number, like -1. -1 could signal that this was the first order, but if I were to run aggregations on this column, the results would get distorted.

So the best choice is to create a new variable to flag these rows as the first order for the customers.

*Since this was not taught in the course yet, another option would be to leave the NaN's there, and make note about what it means. 

In [28]:
# check value counts
flag_count = df_ords['first_order'].value_counts(dropna=False)

In [29]:
flag_count

first_order
False    3214874
True      206209
Name: count, dtype: int64

In [30]:
# export frequency table to excel report
flag_count.to_clipboard()

## Duplicates in the orders df

In [27]:
# check for duplicaates in a new df
df_ords_dups = df_ords[df_ords.duplicated()]

There seems to be no duplicates.

## Export cleaned orders df

In [28]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index=False)